Enforcer rule fails with NullPointerException for a new project

For a user's Maven project that uses linakge-checker-enforcer-rules (default `dependencyManagement:DEPENDENCIES`), the rule's behavior differs depending on locally-installed artifact of the project.

- When a project has already run "mvn clean", LinkageCheckerEnforcerRule checks the class path containing the jar file.

- When a project has not yet run "mvn clean", the rule cannot check the class files in the project.

This is unexpected behavior. Because "mvn validate" lifecycle is before compile, the rule should not touch class files in the project.


Also, the enforcer rule fails with NullPointerException for a new project that has not yet run "mvn install".

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M2:enforce (enforce) on project no-such-method-error-example: Execution enforce of goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M2:enforce failed.: NullPointerException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M2:enforce (enforce) on project no-such-method-error-example: Execution enforce of goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M2:enforce failed.
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:993)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:345)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:191)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution enforce of goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M2:enforce failed.
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:145)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:993)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:345)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:191)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: java.lang.NullPointerException
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:193)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:193)
    at java.util.stream.ReferencePipeline$2$1.accept (ReferencePipeline.java:175)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:193)
    at java.util.Iterator.forEachRemaining (Iterator.java:116)
    at java.util.Spliterators$IteratorSpliterator.forEachRemaining (Spliterators.java:1801)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:481)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:471)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential (ReduceOps.java:708)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.collect (ReferencePipeline.java:499)
    at com.google.cloud.tools.dependencies.enforcer.LinkageCheckerRule.findProjectClasspath (LinkageCheckerRule.java:195)
    at com.google.cloud.tools.dependencies.enforcer.LinkageCheckerRule.execute (LinkageCheckerRule.java:116)
    at org.apache.maven.plugins.enforcer.EnforceMojo.execute (EnforceMojo.java:194)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:134)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:993)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:345)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:191)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[ERROR] 

```

--- a/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
+++ b/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
@@ -26,12 +26,14 @@ import com.google.cloud.tools.opensource.classpath.LinkageCheckReport;
 import com.google.cloud.tools.opensource.classpath.LinkageChecker;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
 import com.google.common.graph.Traverser;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Path;
 import java.util.List;
 import java.util.Objects;
+import java.util.Spliterator;
 import java.util.stream.StreamSupport;
 import javax.annotation.Nonnull;
 import org.apache.maven.RepositoryUtils;
@@ -186,13 +188,26 @@ public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
 
       Traverser<DependencyNode> traverser = Traverser.forTree(DependencyNode::getChildren);
 
-      return StreamSupport.stream(
-              traverser.breadthFirst(resolutionResult.getDependencyGraph()).spliterator(), false)
-          .map(DependencyNode::getArtifact)
-          .filter(Objects::nonNull)
-          .map(Artifact::getFile)
-          .map(File::toPath)
-          .collect(toImmutableList());
+      DependencyNode dependencyGraph = resolutionResult.getDependencyGraph();
+      Iterable<DependencyNode> breadthFirst = traverser.breadthFirst(dependencyGraph);
+      
+      Builder<Path> builder = ImmutableList.builder();
+      for (DependencyNode node : breadthFirst) {
+        // the very first one is the pom.xml where this rule appears
+        Artifact artifact = node.getArtifact();
+        if (artifact != null) { // why is this possible?
+          File file = artifact.getFile();
+          // and this very first one does not have a file; i.e. file == null
+          // but why do we care? perhaps we're assuming there is a jar file in
+          // the classpath but what we really need for this one is a classes directory
+          Path path = file.toPath();
+          builder.add(path);
+        }
+      }
+      ImmutableList<Path> built = builder.build();
+      return built;
+    } catch (NullPointerException e) {
+      throw new EnforcerRuleException("Null pointer " + e.getLocalizedMessage(), e);
     } catch (ComponentLookupException e) {
       throw new EnforcerRuleException("Unable to lookup a component " + e.getLocalizedMessage(), e);
     } catch (DependencyResolutionException e) {

--- a/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
+++ b/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
@@ -24,6 +24,7 @@ import com.google.cloud.tools.opensource.classpath.ClassPathBuilder;
 import com.google.cloud.tools.opensource.classpath.JarLinkageReport;
 import com.google.cloud.tools.opensource.classpath.LinkageCheckReport;
 import com.google.cloud.tools.opensource.classpath.LinkageChecker;
+import com.google.cloud.tools.opensource.dependencies.Artifacts;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
 import com.google.common.graph.Traverser;
@@ -184,8 +185,8 @@ public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
 
       Traverser<DependencyNode> traverser = Traverser.forTree(DependencyNode::getChildren);
 
-      DependencyNode dependencyGraph = resolutionResult.getDependencyGraph();
-      Iterable<DependencyNode> breadthFirst = traverser.breadthFirst(dependencyGraph);
+      Iterable<DependencyNode> breadthFirst =
+          traverser.breadthFirst(resolutionResult.getDependencyGraph());
       
       ImmutableList.Builder<Path> builder = ImmutableList.builder();
       for (DependencyNode node : breadthFirst) {
@@ -196,14 +197,15 @@ public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
           // and this very first one does not have a file; i.e. file == null
           // but why do we care? perhaps we're assuming there is a jar file in
           // the classpath but what we really need for this one is a classes directory
+          if (file == null) {
+            throw new EnforcerRuleException(
+                "Artifact " + Artifacts.toCoordinates(artifact) + " is not associated with a file");
+          }
           Path path = file.toPath();
           builder.add(path);
         }
       }
-      ImmutableList<Path> built = builder.build();
-      return built;
-    } catch (NullPointerException e) {
-      throw new EnforcerRuleException("Null pointer " + e.getLocalizedMessage(), e);
+      return builder.build();
     } catch (ComponentLookupException e) {
       throw new EnforcerRuleException("Unable to lookup a component " + e.getLocalizedMessage(), e);
     } catch (DependencyResolutionException e) {

--- a/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
+++ b/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
@@ -26,15 +26,11 @@ import com.google.cloud.tools.opensource.classpath.LinkageCheckReport;
 import com.google.cloud.tools.opensource.classpath.LinkageChecker;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableList.Builder;
 import com.google.common.graph.Traverser;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Path;
 import java.util.List;
-import java.util.Objects;
-import java.util.Spliterator;
-import java.util.stream.StreamSupport;
 import javax.annotation.Nonnull;
 import org.apache.maven.RepositoryUtils;
 import org.apache.maven.enforcer.rule.api.EnforcerRuleException;
@@ -191,7 +187,7 @@ public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
       DependencyNode dependencyGraph = resolutionResult.getDependencyGraph();
       Iterable<DependencyNode> breadthFirst = traverser.breadthFirst(dependencyGraph);
       
-      Builder<Path> builder = ImmutableList.builder();
+      ImmutableList.Builder<Path> builder = ImmutableList.builder();
       for (DependencyNode node : breadthFirst) {
         // the very first one is the pom.xml where this rule appears
         Artifact artifact = node.getArtifact();

--- a/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
+++ b/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
@@ -183,13 +183,11 @@ public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
       DependencyResolutionResult resolutionResult =
           projectDependenciesResolver.resolve(dependencyResolutionRequest);
 
-      Traverser<DependencyNode> traverser = Traverser.forTree(DependencyNode::getChildren);
-
-      Iterable<DependencyNode> breadthFirst =
-          traverser.breadthFirst(resolutionResult.getDependencyGraph());
+      Iterable<DependencyNode> dependencies = Traverser.forTree(DependencyNode::getChildren)
+          .breadthFirst(resolutionResult.getDependencyGraph());
       
       ImmutableList.Builder<Path> builder = ImmutableList.builder();
-      for (DependencyNode node : breadthFirst) {
+      for (DependencyNode node : dependencies) {
         // the very first one is the pom.xml where this rule appears
         Artifact artifact = node.getArtifact();
         if (artifact != null) { // why is this possible?
@@ -199,7 +197,8 @@ public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
           // the classpath but what we really need for this one is a classes directory
           if (file == null) {
             throw new EnforcerRuleException(
-                "Artifact " + Artifacts.toCoordinates(artifact) + " is not associated with a file");
+                "Artifact " + Artifacts.toCoordinates(artifact) + " is not associated with a file."
+                    + " The linkage checker enforcer rule should be bound to the verify phase.");
           }
           Path path = file.toPath();
           builder.add(path);

